# Домашнее задание к лекции "Статистика. Практика"

## Задание 1

Вернемся к [набору данных о видеоиграх](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/vgsales.csv).

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?  
2) Критикам нравятся больше игры на PC или на PS4?  
3) Критикам больше нравятся стрелялки или стратегии?  

Для каждого вопроса:
- сформулируйте нулевую и альтернативную гипотезы;
- выберите пороговый уровень статистической значимости;
- опишите полученные результаты статистического теста.

In [1]:
import pandas as pd

In [2]:
game_data = pd.read_csv('vgsales.csv')
game_data

Name Platform  Year_of_Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16714  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16715               LMA Manager 2007     X360           2006.0        Sports   
16716        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16717               Spirits & Spells      GBA           2003.0      Platform   
16718            Winning Post 8 2016      PSV           2016.0    Simulation   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0          Nintendo     41.36     28.96      3.77         8.45         82.53   
1          Nintendo     29.08      3.58      6.81         0.77         40.24   
2          Nintendo     15.68     12.76      3.79         3.29         35.52   
3          Nintendo     15.61     10.93      3.28         2.95         32.77   
4          Nintendo     11.27      8.89     10.22         1.00         31.37   
...             ...       ...       ...       ...          ...           ...   
16714    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   
16715   Codemasters      0.00      0.01      0.00         0.00          0.01   
16716  Idea Factory      0.00      0.00      0.01         0.00          0.01   
16717       Wanadoo      0.01      0.00      0.00         0.00          0.01   
16718    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   

       Critic_Score  Critic_Count  User_Score  User_Count Developer Rating  
0              76.0          51.0         8.0       322.0  Nintendo      E  
1               NaN           NaN         NaN         NaN       NaN    NaN  
2              82.0          73.0         8.3       709.0  Nintendo      E  
3              80.0          73.0         8.0       192.0  Nintendo      E  
4               NaN           NaN         NaN         NaN       NaN    NaN  
...             ...           ...         ...         ...       ...    ...  
16714           NaN           NaN         NaN         NaN       NaN    NaN  
16715           NaN           NaN         NaN         NaN       NaN    NaN  
16716           NaN           NaN         NaN         NaN       NaN    NaN  
16717           NaN           NaN         NaN         NaN       NaN    NaN  
16718           NaN           NaN         NaN         NaN       NaN    NaN  

[16719 rows x 16 columns]

### Как критики относятся к спортивным играм?

Н0 - средний рейтинг критиков для спортивных игр >= 70 (положительно относятся) <br>
H1 - средний рейтинг < 70 <br>
Уровень значимости α примем за 5%

In [3]:
from scipy import stats
alpha = 0.05
result = stats.ttest_1samp(game_data.loc[game_data.Genre=='Sports'].Critic_Score.dropna(), 70, alternative='less')

# средняя оценка спортивных игр
print(game_data.loc[game_data.Genre=='Sports'].Critic_Score.mean())
# результат t-теста
print(result)

if result.pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, среднее меньше 70 (критики негативно относятся к спортивным играм)')
else:
    print('Не отвергаем нулевую нулевую гипотезу, критики положительно относятся к спортивным играм')

71.96817420435511
Ttest_1sampResult(statistic=4.900329475499578, pvalue=0.99999945572554)
Не отвергаем нулевую нулевую гипотезу, критики положительно относятся к спортивным играм


### Критикам нравятся больше игры на PC или на PS4?

Н0 - разницы нет (средняя оценка по выборкам одинакова) (критики одинаково относятся к играм на PC и PS4) <br>
H1 - есть статистически значимая разница между средними в выборках <br>
Уровень значимости α примем за 5%

In [4]:
result = stats.ttest_ind(game_data.loc[game_data.Platform=='PC'].Critic_Score.dropna(),  
                         game_data.loc[game_data.Platform=='PS4'].Critic_Score.dropna())

# средние оценки игр на РС и PS4
print(game_data.loc[game_data.Platform=='PC'].Critic_Score.mean(), 
      game_data.loc[game_data.Platform=='PS4'].Critic_Score.mean())
# результаты t-теста
print(result)  

if result.pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, критикам больше нравятся игры на {}'.format('PC' if result.statistic > 0 else 'PS4'))
else:
    print('Не отвергаем нулевую нулевую гипотезу, критики одинаково относятся к играм на РС и PS4')

75.92867132867133 72.09126984126983
Ttest_indResult(statistic=4.544112457929973, pvalue=6.215797831285281e-06)
Отвергаем нулевую гипотезу, критикам больше нравятся игры на PC


### Критикам больше нравятся стрелялки или стратегии?

Н0 - разницы нет (средняя оценка по выборкам одинакова) (критики одинаково относятся к шутерам и стратегиям) <br>
H1 - есть статистически значимая разница между средними в выборках <br>
Уровень значимости α примем за 5%

In [5]:
result = stats.ttest_ind(game_data.loc[game_data.Genre=='Shooter'].Critic_Score.dropna(),  
                         game_data.loc[game_data.Genre=='Strategy'].Critic_Score.dropna())

# средние оценки шутеров и стратегий
print(game_data.loc[game_data.Genre=='Shooter'].Critic_Score.mean(), 
      game_data.loc[game_data.Genre=='Strategy'].Critic_Score.mean())
# результаты t-теста
print(result)  

if result.pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, критикам больше нравятся {}'.format('шутеры' if result.statistic > 0 else 'стратегии'))
else:
    print('Не отвергаем нулевую нулевую гипотезу, критики одинаково относятся к шутерам и стратегиям')

70.18114406779661 72.08609271523179
Ttest_indResult(statistic=-2.06670263948504, pvalue=0.03896871303135602)
Отвергаем нулевую гипотезу, критикам больше нравятся стратегии


## Задание 2

Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные [здесь](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/spam.csv)) по признаку спама. Для этого:

1) Приведите весь текст к нижнему регистру;  
2) Удалите мусорные символы;  
3) Удалите стоп-слова;  
4) Привидите все слова к нормальной форме;  
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:  

```
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)
```

Можете поэкспериментировать с параметрами [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html);  <br>
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите `random_state=42`. Используйте [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html); <br>
7) Постройте модель [логистической регрессии](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html), укажите `random_state=42`, оцените ее точность на тестовых данных;  
8) Опишите результаты при помощи [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html?highlight=confusion_matrix#sklearn.metrics.confusion_matrix);  <br>
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [6]:
spam_data = pd.read_csv('spam.csv')
spam_data

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [7]:
import re

spam_data.Message = spam_data.Message.apply(lambda x: x.lower())   # приводим к нижнему регистру
spam_data.Message = spam_data.Message.apply(lambda x: re.sub('[\W_]+',' ', x))  # убираем знаки препинания

In [8]:
# убираем стоп-слова
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))
spam_data.Message = spam_data.Message.apply(lambda x: [word for word in x.split(' ') if word not in stopwords_set])

In [9]:
# лемматизируем слова
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
spam_data.Message = spam_data.Message.apply(lambda x: ' '.join([wordnet_lemmatizer.lemmatize(word) for word in x]))

In [10]:
# получаем матрицу TD-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(spam_data.Message.tolist())
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)
tfidf_matrix

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8048 columns]

In [11]:
# разбиваем данные на тренировочную и тестовую выборку
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, spam_data.Category, test_size=0.30, random_state=42)


In [15]:
# создаём и обучаем модель
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score

lda = LinearDiscriminantAnalysis()
lda.fit(X_train, y_train)
lda.predict(X_test)
accuracy_score(y_test, lda.predict(X_test))

0.9677033492822966

Получаем хороший показатель точности - 0.967.

In [16]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, lda.predict(X_test))

array([[1445,    3],
       [  51,  173]], dtype=int64)

Из матрицы ошибок видим, что 1618 (1445+173) из 1672 (размер тестовой выборки) значений были предсказаны корректно. 54 значения получены неверно. Представим их в таблице:

In [17]:
mismatched = spam_data.loc[y_test.index, :].reset_index().join(pd.DataFrame({'Predicted': lda.predict(X_test)}))
mismatched = mismatched.loc[mismatched.Category != mismatched.Predicted]
mismatched[['Message', 'Category', 'Predicted']]

Message Category Predicted
22                   err cud going 8pm got way contact       ham      spam
40    reminder downloaded content already paid goto ...     spam       ham
47    guess somebody know secretly fancy wanna find ...     spam       ham
74    oh god found number glad text back xafter msg ...     spam       ham
84    next amazing xxx picsfree1 video sent enjoy on...     spam       ham
160   babe u want dont u baby im nasty thing 4 filth...     spam       ham
166   want cock hubby away need real man 2 satisfy t...     spam       ham
194   hello going village pub 8 either come accordin...      ham      spam
234   hello darling today would love chat dont tell ...     spam       ham
260   guess somebody know secretly fancy wanna find ...     spam       ham
272   realize 40 year thousand old lady running arou...     spam       ham
324   bloomberg message center 447797706009 wait app...     spam       ham
344   400 xmas reward waiting computer randomly pick...     spam       ham
393   chance win free bluetooth headset simply reply...     spam       ham
408   tddnewsletter emc1 co uk game thedailydraw dea...     spam       ham
426   urgent ur 500 guaranteed award still unclaimed...     spam       ham
461   taking part mobile survey yesterday 500 text 2...     spam       ham
489   2nd time tried contact u u 1450 prize claim ca...     spam       ham
518   u meet ur dream partner soon ur career 2 flyng...     spam       ham
530   link picture sent also use http alto18 co uk w...     spam       ham
541   22 day kick euro2004 u kept date latest news r...     spam       ham
550   bloomberg message center 447797706009 wait app...     spam       ham
561   talk sexy make new friend fall love world disc...     spam       ham
576   urgent 2nd attempt contact u u 1000call 090715...     spam       ham
611   dear u invited xchat final attempt contact u t...     spam       ham
635         sorry missed call let talk time 07090201529     spam       ham
643   hack chat get backdoor entry 121 chat room fra...     spam       ham
678   mobile club choose top quality item mobile 7cf...     spam       ham
681   married local woman looking discreet action 5 ...     spam       ham
731   400 xmas reward waiting computer randomly pick...     spam       ham
773   new 2 club dont fink met yet b gr8 2 c u pleas...     spam       ham
795   burger king wanna play footy top stadium get 2...     spam       ham
807                                  yar say got error       ham      spam
927                  private 2003 account statement 078     spam       ham
1002  bear pic nick tom pete dick fact type try gay ...     spam       ham
1081  email alertfrom jeri stewartsize 2kbsubject lo...     spam       ham
1107             1 new message please call 08712400200      spam       ham
1132  call 09090900040 listen extreme dirty live cha...     spam       ham
1175                                 88066 lost 12 help     spam       ham
1228  text pas 69669 collect polyphonic ringtones no...     spam       ham
1269  free msg get gnarl barkley crazy ringtone tota...     spam       ham
1279   forwarded 448712404000 please call 0871240400...     spam       ham
1294  thanks vote sing along star karaoke mobile fre...     spam       ham
1302            hear new divorce barbie come ken stuff      spam       ham
1329  current leading bid 151 pause auction send cus...     spam       ham
1362  summer finally fancy chat flirt sexy single yr...     spam       ham
1383  lookatme thanks purchase video clip lookatme c...     spam       ham
1407  boltblue tone 150p reply poly mono eg poly3 1 ...     spam       ham
1475  hungry gay guy feeling hungry 4 call 087187305...     spam       ham
1495  cmon babe make horny turn txt fantasy babe im ...     spam       ham
1501  discount code rp176781 stop message reply stop...     spam       ham
1509  unique enough find 30th august www areyouuniqu...     spam       ham
1572  free msg single find partner area 1000 real pe...     spam       ha

Из матрицы ошибок и из таблицы выше видим, что всего у нас 54 неверно классифицированных сообщений, из них 51 ложноотрицательных результатов (ошибок 2-ого рода): модель определила сообщение как не спам, хотя это был спам, и 3 ложноположительных.

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам?
Прежде чем задать вопрос необходимо попробовать найти ответ самому в интернете. Навык самостоятельного поиска информации – один из важнейших, и каждый практикующий специалист любого уровня это делает каждый день.

Любой вопрос должен быть сформулирован по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности, прикрепляйте к вопросу скриншоты, либо ссылки на код. Оставляйте только проблемный и воспроизводимый участок кода, все решение выкладывать не допускается.
